### ISSUE TITLE
Auto port / stream id assignment

### ISSUE THREAD

**Issue by lbradstreet** [MichaelDrogalis/onyx#123]
This issue is somewhat related to https://github.com/MichaelDrogalis/onyx/issues/122.

I started down a track where I moved the port and stream assignment into apply-log-entries. I figured the observer could calculate unused stream ids (and the port, if it is not fixed) in `:notify-join-cluster`, and then the vpeer could actually start the subscriptions as a side-effect to the `:accept-join-cluster`. I think this approach would work, and I'd be interested in hear if not (for my own understanding!).

That said, a simpler approach may be leveraging the peer group in #122 and using an atom of current assignments. Assignments would be added when vpeers startup and removed when vpeers shutdown. We only need to make sure there are no port/stream-id collisions on the current instance, and the current approach i.e. publishing the port/stream-ids in `:prepare-join-cluster` should still work. One downside with this approach is what may happen when a vpeer does not have a chance to shut itself down properly. I guess if this atom used by the PeerGroup is included as part of the state, then we could automatically deregister the peer from this atom on `:abort-join` or `:leave-cluster` (the peer would only exist in this atom if the vpeer is running on the machine).

Thoughts?

**Comment by MichaelDrogalis** [comment]
The first approach should work. Here's some background...

If a peer advertises a connection to receive messages, and it doesn't have the channels set up to actually *do* anything with those messages - it will drop them or buffer them. The channels are set up inside the task lifecycle. This is why we have the two peer states `:warming-up` and `:active`. Warming up denotes that it's still constructing the supporting channels to receive messages, and active means "Okay, everything is ready. I can receive messages." I think this approach can certainly work since we already have the notion of warming-up built in.

I sort of like the first approach best - each time a peer registers, we can look at all other peers on the same host, look for the highest port, and pick the next port. That let's us reliability demarcate a sequences of channels to be used for N peers (e.g. 40000 - (40000 + N)).

**Comment by lbradstreet** [comment]
OK, cool. 

Just to be clear with my thinking, the vpeer sends out prepare-join with its ip to start, then it catches up with the log. In the observer's prepare-join reactions, it calculates the resources it wants to reserve (steam ids, ports, etc). It can do so because it knows which vpeers are on which machines thanks to the ip advertised in prepare-join. These are returned in the notify-join data. 

Then the channel/port/streams/etc are setup by the vpeer when it sees notify-join for itself, which point it says it's ready for work (alternatively in accept-join? Maybe this is too late. I found the difference between accept join and notify join a little confusing when I was working on the dashboard). 

That port / stream allocation plan is what I was thinking. We can pass the required information into a messaging multimethod and let it decide. That way implementation details (e.g streams vs ports) can be left to the messaging. 

I'm happy to finish the work on this one myself. Digging into the log some more has been good.

**Comment by MichaelDrogalis** [comment]
First paragraph was spot on. Second paragraph isn't a problem - it's not too late. Accept and notify are a little annoying, but there needs to be a distinction to ensure that the peer safely joins.

**Comment by lbradstreet** [comment]
This is basically done, though I backed out the multi-message assignment as it required the second peer to abort join and I didn't fully understand the strategy at the time. I may give it another go when I get some time. 

Unfortunately I needed to pass the messenger to apply-log-entry. However, the only function that is called is pure.

**Comment by MichaelDrogalis** [comment]
A lot of files had to change because of a single log entry needing the Messenger. I took a look at the pull request - we might have an easier time doing what we did with the scheduler. That is, we supply the messenger type as a keyword in the replica (`{:messenger-type :aeron}`) and use a multimethod inside the apply-log-entry body to figure out how to pick the correct port. Sorry I didn't notice this earlier, I thought the Messenger was being used inside `reactions` for some reason, which already had the state atom. Does this seem reasonable?
